joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h

```diff
@@ -811,6 +688,151 @@ setHoldPosition(const ros::Time& time, RealtimeGoalHandlePtr gh)
   curr_trajectory_box_.set(hold_trajectory_ptr_);
 }
 
+template <class SegmentImpl, class HardwareInterface>
+void JointTrajectoryController<SegmentImpl, HardwareInterface>::
+updateStateAndTimeData(const ros::Time& time, const ros::Duration& period, TimeData& time_data)
```

Comment:
Why does this function do two things (even indicated by the "and" in the name)? For the sake of clarity I suggest to create functions with one responsibility only. (For more information about the single responsibility see also "Clean Code" by Robert C. Martin Chapter 3  "Functions")

I suggest separating the getting of the trajectory and the time update from this function and just make this function about the state updates and checks.